File: apps/mobile/modules/app-icon/android/src/main/java/expo/modules/appicon/AppIconModule.kt
package expo.modules.appicon

import android.content.ComponentName
import android.content.pm.PackageManager
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private const val defaultIconId = "crimson-bars"
private val aliasSuffixes = linkedMapOf(
  "dark-bars" to "AppIconDarkBars",
  "gold-glow" to "AppIconGoldGlow",
  defaultIconId to "AppIconCrimsonBars"
)

class AppIconModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AppIcon")

    AsyncFunction("supportsAlternateIcons") { true }

    AsyncFunction("getAppIcon") {
      val packageManager = context.packageManager
      aliasSuffixes.entries.firstOrNull { (iconId, suffix) ->
        when (packageManager.getComponentEnabledSetting(component(suffix))) {
          PackageManager.COMPONENT_ENABLED_STATE_ENABLED -> true
          PackageManager.COMPONENT_ENABLED_STATE_DEFAULT -> iconId == defaultIconId
          else -> false
        }
      }?.key ?: defaultIconId
    }

    AsyncFunction("setAppIcon") { iconId: String ->
      val targetSuffix = aliasSuffixes[iconId] ?: throw InvalidAppIconException(iconId)
      val packageManager = context.packageManager
      val previousStates = aliasSuffixes.values.associateWith { suffix ->
        packageManager.getComponentEnabledSetting(component(suffix))
      }

      try {
        packageManager.setComponentEnabledSetting(
          component(targetSuffix),
          PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
          PackageManager.DONT_KILL_APP
        )
        aliasSuffixes.values.filter { it != targetSuffix }.forEach { suffix ->
          packageManager.setComponentEnabledSetting(
            component(suffix),
            PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
            PackageManager.DONT_KILL_APP
          )
        }
      } catch (error: Exception) {
        previousStates.forEach { (suffix, state) ->
          packageManager.setComponentEnabledSetting(component(suffix), state, PackageManager.DONT_KILL_APP)
        }
        throw error
      }
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext) {
      "React Application Context is null"
    }

  private fun component(suffix: String) =
    ComponentName(context.packageName, "${context.packageName}.$suffix")
}

private class InvalidAppIconException(iconId: String) :
  CodedException("ERR_INVALID_APP_ICON", "未知应用图标：$iconId", null)

File: apps/mobile/modules/app-icon/ios/AppIconModule.swift
import ExpoModulesCore
import UIKit

private let defaultIconId = "crimson-glass"
private let alternateIconNames = [
  "dark-bars": "AppIconDarkBars",
  "gold-glow": "AppIconGoldGlow",
  "crimson-bars": "AppIconCrimsonBars"
]

public class AppIconModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AppIcon")

    AsyncFunction("supportsAlternateIcons") {
      UIApplication.shared.supportsAlternateIcons
    }.runOnQueue(.main)

    AsyncFunction("getAppIcon") { () -> String in
      guard let alternateName = UIApplication.shared.alternateIconName else {
        return defaultIconId
      }
      return alternateIconNames.first(where: { $0.value == alternateName })?.key ?? defaultIconId
    }.runOnQueue(.main)

    // Swift 5 的同步 closure 里不能用 withCheckedThrowingContinuation（async-only API）；
    // Expo 57 的 ConcurrentFunction 重载用了 sending 关键字，Swift 5 编译器不识别。
    // 所以用 DispatchGroup 在 Expo 后台线程上阻塞等主线程回调完成。
    // 刻意**不**加 .runOnQueue(.main)：否则 semaphore.wait() 会阻塞主线程、死锁。
    AsyncFunction("setAppIcon") { (iconId: String) in
      guard iconId == defaultIconId || alternateIconNames[iconId] != nil else {
        throw InvalidAppIconException(iconId)
      }
      guard UIApplication.shared.supportsAlternateIcons else {
        throw UnsupportedAppIconException()
      }

      let iconName = iconId == defaultIconId ? nil : alternateIconNames[iconId]
      let group = DispatchGroup()
      group.enter()
      var resultError: Error?

      DispatchQueue.main.async {
        UIApplication.shared.setAlternateIconName(iconName) { error in
          resultError = error
          group.leave()
        }
      }

      group.wait()

      if let error = resultError {
        throw error
      }
    }
  }
}

private class InvalidAppIconException: GenericException<String>, @unchecked Sendable {
  override var reason: String {
    "未知应用图标：\(param)"
  }
}

private class UnsupportedAppIconException: Exception, @unchecked Sendable {
  override var reason: String {
    "当前设备不支持切换应用图标"
  }
}
